### DISCUSSION TITLE
How to handle static generated inputs with relay plugin?

### DISCUSSION THREAD

**Discussion by lakhanmandloi** [hayes/pothos#1251] [Q&A] [answered]
Hi there,

## Problem statement

I would like to know the right approach of using relay plugin with static generated input types.

## Description

I generate Input types using [static generation](https://pothos-graphql.dev/docs/plugins/prisma/prisma-utils#static-generator).

For the `User` model:

```prisma
model User {
  id Int @id @default(autoincrement())
  name String
  email String @unique
}
```

The generated `UserUniqueFilter` input type is:

```ts
export const UserUniqueFilter = builder.prismaWhereUnique("User", {
	name: "UserUniqueFilter",
	fields: () => ({
		id: "Int",
		email: "String",
	}),
});
```

to use relay plugin for pagination.

```ts
builder.prismaNode("User", {
	name: "User",
	id: { field: "id" },
	fields: (t) => ({
		name: t.exposeString("name", {
			description: "Name of the user",
			nullable: false,
		}),
		email: t.exposeString("email", {
			description: "Email of the user",
			nullable: false,
		}),
	}),
});
```

When querying user/s, the returned data is:

```json
{
	"data": {
		"users": {
			"pageInfo": {
				"endCursor": "R1BDOk46Mjk=",
				"hasNextPage": false,
				"hasPreviousPage": false,
				"startCursor": "R1BDOk46MQ=="
			},
			"edges": [
				{
					"cursor": "R1BDOk46MQ==",
					"node": {
						"name": "John Doe",
						"email": "john@doe.com",
						"id": "TGFuZ3VhZ2U6MQ=="
					}
				},
				{
					"cursor": "R1BDOk46Mg==",
					"node": {
						"name": "Jane Doe",
						"email": "jane@doe.com",
						"id": "TGFuZ3VhZ2U6Mg=="
					}
				}
			]
		}
	}
}
```

the returned id is global id / string, but the `UserUniqueFilter` input type expects `Int` type.

What should I do now? What can be the right approach to handle this?

1. Should I change static generation process to generate the `UserUniqueFilter` input type to accept `id` as `string` type and decode it in resolve ?
2. Should I add a new field in `prismaNode` to expose actual id with different name and use it as `id` in `UserUniqueFilter` input type?
3. Any other approach?

**Comment by hayes** [comment, marked as answer]
I think this will be up to the app, and there isn't an easy answer.

The prisma-utils plugin doesn't have anything built in to handle this today.

I think exposing the raw ID as an Int field might be the best option for you here.  When generating your schema from the DB, using the values as they exist in the DB is probably the right call.   I think if you want to have a more tailored experience, you can customize the generator code to generate IDs as globalID inputs, and write a plugin that automatically decodes them to the right type, but doing this correctly isn't trivial.

These kinds of edge cases are why prisma-utils and generators are marked as experimental.  They work well enough for the basic cases, and aren't likely to have a lot of bugs, but they don't handle every case well.

If you are interested in contributing a good solution here, I am definitely happy to help out, and advise, but I don't know any good existing solutions to this specific problem
